<a href="https://colab.research.google.com/github/schauppi/Udemy_TensorFlow_Developer/blob/schauppi/Abschnitt_2_Deep_Learning_and_TensorFlow_Fundamentals/00_tensorflow_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import tensorflow as tf
print(tf.__version__)

2.5.0


Get Information about function - click in function an press shift + command + space

In [2]:
#create tensor with tf.constant()
scalar = tf.constant(7)
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=7>

In [3]:
#Check the number of dimensions of a tensor (ndim = number of dimensions)
scalar.ndim

0

In [4]:
#create a vector
vector = tf.constant([10, 10])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([10, 10], dtype=int32)>

In [5]:
#check the dimensions of our vector
vector.ndim

1

In [6]:
#create a matrix (has more than 1 dimension)
matrix = tf.constant([[10, 7], [7, 10]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7, 10]], dtype=int32)>

In [7]:
matrix.ndim

2

Higher number in the Datatype means more accurate (more exact) saving of the number - which means it needs more disk space on the computer. float16 needs less space on disk than float32


In [8]:
#create another matrix
another_matrix = tf.constant([[10., 7.], 
                              [3., 2.], 
                              [8., 9.]], dtype=tf.float16) #specify datatype
another_matrix

<tf.Tensor: shape=(3, 2), dtype=float16, numpy=
array([[10.,  7.],
       [ 3.,  2.],
       [ 8.,  9.]], dtype=float16)>

In [9]:
#Number of dimension of another_matrix
another_matrix.ndim

2

In [10]:
#create a tensor
tensor = tf.constant([[[1, 2, 3], 
                       [4, 5, 6]], 
                     [[7, 8, 9], 
                       [10, 11, 12]],
                     [[13, 14, 15], 
                       [16, 17, 18]]])
tensor

<tf.Tensor: shape=(3, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18]]], dtype=int32)>

In [11]:
tensor.ndim

3

Scalar - 0 Dimensions <br>
Vector - 1 Dimensions <br>
Matrix - 2 Dimensions <br>
Tensor - 3 Dimensions<br>

New Code Cell: <br>
Command - M - B

###Creating tensors with tf.Variable

In [12]:
#Create the same tensor with tf.Variable() as above
changeable_tensor = tf.Variable([10, 7])
unchangeable_tensor = tf.constant([10, 7])
changeable_tensor, unchangeable_tensor

(<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([10,  7], dtype=int32)>,
 <tf.Tensor: shape=(2,), dtype=int32, numpy=array([10,  7], dtype=int32)>)

In [13]:
#Try to changhe one of the elements in changeable tensor
#changeable_tensor[0] = 7
#changeable_tensor

In [14]:
#try .assign()
changeable_tensor[0].assign(7)
changeable_tensor

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([7, 7], dtype=int32)>

In [15]:
#change unchangeable tensor
#unchangeable_tensor[0].assign(7)
#unchangeable_tensor

###Creating random Tensors
Random tensors are tensors of some abitrary size which contain random numbers

In [16]:
#Create two random (but the same) tensors
random_1 = tf.random.Generator.from_seed(42) #set seed for reproduceibility
random_1 = random_1.normal(shape=(3, 2))
random_1

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702],
       [ 0.07595026, -1.2573844 ],
       [-0.23193765, -1.8107855 ]], dtype=float32)>

In [17]:
random_2 = tf.random.Generator.from_seed(42)
random_2 = random_2.normal(shape=(3, 2))
random_2

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702],
       [ 0.07595026, -1.2573844 ],
       [-0.23193765, -1.8107855 ]], dtype=float32)>

In [18]:
#Are the tensors equal?
random_1, random_2, random_1 == random_2

(<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193765, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193765, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=bool, numpy=
 array([[ True,  True],
        [ True,  True],
        [ True,  True]])>)

###Shuffle the order of elements in tensor

In [19]:
#Shuffle a tensor (valuable for when you want to shuffle your data so the inherent order wont affect training)
not_shuffled = tf.constant([[10, 7], 
                            [3, 4], 
                            [2, 5]])
not_shuffled

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [20]:
not_shuffled.ndim

2

In [21]:
#shuffled along first dimension
tf.random.shuffle(not_shuffled)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [22]:
#shuffle our non-shuffled tensor
tf.random.shuffle(not_shuffled, seed=42)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [ 3,  4],
       [10,  7]], dtype=int32)>

In [23]:
not_shuffled

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

If we want out shuffled tensors to be in the same order, we´ve got use the global level random seed as well as the operation level random seed:

> Rule 4: If both the global and the operation seed are set: Both seeds are used in conjunction to determine the random sequence.

In [24]:
tf.random.set_seed(42) #global level random seed
tf.random.shuffle(not_shuffled, seed=42) #operation level random seed

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

###Other ways to make tensors

In [25]:
#Create a tensor of all ones
tf.ones([10, 7])

<tf.Tensor: shape=(10, 7), dtype=float32, numpy=
array([[1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.]], dtype=float32)>

In [26]:
#Create a tensor of all zeros
tf.zeros([10, 7])

<tf.Tensor: shape=(10, 7), dtype=float32, numpy=
array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]], dtype=float32)>

###Turn numpy arrays into tensors

The main difference between NumPy arrays and Tensorfs is taht tensors can be run on a GPU

In [27]:
import numpy as np
numpy_A = np.arange(1, 25, dtype=np.int32) #create a NumPy array between 1 and 25
numpy_A

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)

In [28]:
A = tf.constant(numpy_A, shape=(2, 3, 4))
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]], dtype=int32)>

In [29]:
B = tf.constant(numpy_A)
B

<tf.Tensor: shape=(24,), dtype=int32, numpy=
array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)>

###Getting informations from tensors

* Shape
* Rank
* Axis or dimensions
* Size

In [30]:
#Create a rank 4 tensor (4 dimensions)
rank_4_tensor = tf.zeros([2, 3, 4, 5])
rank_4_tensor

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [31]:
rank_4_tensor[0]

<tf.Tensor: shape=(3, 4, 5), dtype=float32, numpy=
array([[[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]]], dtype=float32)>

In [32]:
rank_4_tensor.shape, rank_4_tensor.ndim, tf.size(rank_4_tensor)

(TensorShape([2, 3, 4, 5]), 4, <tf.Tensor: shape=(), dtype=int32, numpy=120>)

In [33]:
#Get various attributes of our tensor
print("Datatype of every element:", rank_4_tensor.dtype)
print("Number of dimensions (rank):", rank_4_tensor.ndim)
print("Shape of tensor:", rank_4_tensor.shape)
print("Elements along the 0 axis:", rank_4_tensor.shape[0])
print("Elements along the last axis:", rank_4_tensor.shape[-1])
print("Total number of elements in our tensor:", tf.size(rank_4_tensor))
print("Total number of elements in our tensor:", tf.size(rank_4_tensor).numpy())

Datatype of every element: <dtype: 'float32'>
Number of dimensions (rank): 4
Shape of tensor: (2, 3, 4, 5)
Elements along the 0 axis: 2
Elements along the last axis: 5
Total number of elements in our tensor: tf.Tensor(120, shape=(), dtype=int32)
Total number of elements in our tensor: 120


###Indexing tensors

Tensors can be indexed just like Python lists

In [34]:
#Get the first 2 elements of each dimension
rank_4_tensor[:2, :2, :2, :2]

<tf.Tensor: shape=(2, 2, 2, 2), dtype=float32, numpy=
array([[[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]],


       [[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]], dtype=float32)>

In [35]:
#Get the first element from each dimension from each index except for the final one
rank_4_tensor[:1, :1, :1, :]

<tf.Tensor: shape=(1, 1, 1, 5), dtype=float32, numpy=array([[[[0., 0., 0., 0., 0.]]]], dtype=float32)>

In [36]:
rank_4_tensor.shape

TensorShape([2, 3, 4, 5])

In [37]:
#Create a rank 2 tensor (2 dimensions)
rank_2_tensor = tf.constant([[10,7],
                            [3, 4]])

rank_2_tensor.shape, rank_4_tensor.ndim

(TensorShape([2, 2]), 4)

In [38]:
#Get the last item of each row of our rank 2 tensor
rank_2_tensor[:, -1]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([7, 4], dtype=int32)>

In [39]:
#Add in extra dimension to our rank 2 tensor
rank_3_tensor = rank_2_tensor[..., tf.newaxis] #add new axis on the end
rank_3_tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

In [40]:
#Alternative to tf.newaxis
tf.expand_dims(rank_2_tensor, axis=-1) #-1 means expand last axis


<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

In [41]:
tf.expand_dims(rank_2_tensor, axis=0) #expand the 0-axis

<tf.Tensor: shape=(1, 2, 2), dtype=int32, numpy=
array([[[10,  7],
        [ 3,  4]]], dtype=int32)>

In [42]:
tf.expand_dims(rank_2_tensor, axis=1)

<tf.Tensor: shape=(2, 1, 2), dtype=int32, numpy=
array([[[10,  7]],

       [[ 3,  4]]], dtype=int32)>

In [43]:
rank_2_tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

###Manipulating tensors (tensor operations)

**Basic operations

+, -, *, /

In [44]:
#You can add values to a tensor using the addition operator
tensor = tf.constant([[10, 7],
                      [3,4]])

tensor + 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[20, 17],
       [13, 14]], dtype=int32)>

In [45]:
#Original tensor is unchanged
tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

In [46]:
#Multiplication
tensor * 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  70],
       [ 30,  40]], dtype=int32)>

In [47]:
#Subtraction
tensor -10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 0, -3],
       [-7, -6]], dtype=int32)>

In [48]:
#We can use the tensorflow built in function too
#Faster than the other one
tf.multiply(tensor, 10)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  70],
       [ 30,  40]], dtype=int32)>

###Matrix multiplication

Most common operations in ML

In [49]:
#Matrix multiplication in tensorflow
print(tensor)
tf.matmul(tensor, tensor)

tf.Tensor(
[[10  7]
 [ 3  4]], shape=(2, 2), dtype=int32)


<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [50]:
#Matrix multiplication with Python operator "@"
tensor @ tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [51]:
tensor_a = tf.constant([[1,2,5], 
              [7,2,1],
              [3,3,3]])

In [52]:
tensor_b = tf.constant([[3, 5],
                        [6, 7],
                        [1, 8]])

In [53]:
tf.matmul(tensor_a,tensor_b)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[20, 59],
       [34, 57],
       [30, 60]], dtype=int32)>

In [54]:
print(tensor_a.shape)
print(tensor_b.shape)

(3, 3)
(3, 2)


In [55]:
tensor_c = tf.constant([[1,2],
                        [3,4],
                        [5,6]])

In [56]:
#change shape of tensor_c
tensor_d = tf.reshape(tensor_c, shape=(2,3))
tensor_d

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[1, 2, 3],
       [4, 5, 6]], dtype=int32)>

In [57]:
tf.matmul(tensor_c, tensor_d)

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 9, 12, 15],
       [19, 26, 33],
       [29, 40, 51]], dtype=int32)>

In [58]:
#matrix multiplication is possible - right dimensions
tensor_c.shape, tensor_d.shape

(TensorShape([3, 2]), TensorShape([2, 3]))

In [59]:
#Can do the same with transpose
#Flips rows and columns
tf.transpose(tensor_c)

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[1, 3, 5],
       [2, 4, 6]], dtype=int32)>

In [60]:
#Try matrix multiplication with transpose
tf.matmul(tf.transpose(tensor_c), tf.transpose(tensor_d))

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[22, 49],
       [28, 64]], dtype=int32)>

###The dot product
Matrix multiplication is also referred as the dot product. <br>
You can perform matrix multiplication with: <br>
* `tf.matmul()`
* `tf.tensordot()`

In [61]:
#Perform the dot product on tensor_c and tensor_d
print(tensor_c.shape)
print(tensor_d.shape)

(3, 2)
(2, 3)


In [62]:
tf.tensordot(tensor_c, tensor_d, axes=1)

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 9, 12, 15],
       [19, 26, 33],
       [29, 40, 51]], dtype=int32)>

In [63]:
#Perform matrix multiplication bewteen tensor_c, tensor_d (transposed)
tf.matmul(tf.transpose(tensor_c), tf.transpose(tensor_d))

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[22, 49],
       [28, 64]], dtype=int32)>

In [64]:
#Perform matrix multiplication bewteen tensor_c, tensor_d (reshaped)
tf.matmul(tf.reshape(tensor_c, shape=(2,3)), tf.reshape(tensor_d, shape=(3,2)))

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[22, 28],
       [49, 64]], dtype=int32)>

In [65]:
#Check the vakues of tensor_c, reshaped tensor_c and transposed tensor_c
print("Normal: ", "\n", tensor_c) # "\n" is for new line
print("reshaped: ", "\n", tf.reshape(tensor_c, shape=(2,3)))
print("transposed: ", "\n", tf.transpose(tensor_c))

Normal:  
 tf.Tensor(
[[1 2]
 [3 4]
 [5 6]], shape=(3, 2), dtype=int32)
reshaped:  
 tf.Tensor(
[[1 2 3]
 [4 5 6]], shape=(2, 3), dtype=int32)
transposed:  
 tf.Tensor(
[[1 3 5]
 [2 4 6]], shape=(2, 3), dtype=int32)


###Transpose the tensors when one of the axis doesn´t line up rather than reshaping it!!!!!!!! 

###Change the datatype of a tensor
Reducing the amount of used memory per tensor

In [66]:
#Create a new tensor with default datatype (float32)
B = tf.constant([1.7, 7.4])
B.dtype

tf.float32

In [67]:
C = tf.constant([7,10])
C.dtype

tf.int32

In [68]:
#Change from float32 to float16 - reduced precision - more computation speed
B = tf.cast(B, dtype=tf.float16)
B.dtype

tf.float16

In [69]:
C = tf.cast(B, dtype=tf.int16)
C.dtype

tf.int16

###Aggregating tensor

Aggregating tensors = condensing them from multiple values down to a smaller amount of values

In [70]:
#Get the absolute values
D = tf.constant([-7, -10])
D

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ -7, -10], dtype=int32)>

In [71]:
#Get the absolute values
#turn negative into positive values
tf.abs(D)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10], dtype=int32)>

Let#s go through the following forms of aggregation:
* Get the minimumü
* Get the maximum
* Get the mean of tensor
* Get the sum of a tensor

In [72]:
#Creating a random tensor with values between 0 and 100 of size 50
E = tf.constant(np.random.randint(0, 100, size=50))
E

<tf.Tensor: shape=(50,), dtype=int64, numpy=
array([49, 82, 72, 98,  2, 66, 83, 61, 61, 53, 12, 37, 14, 15, 79, 17, 52,
       71, 33, 76, 87, 42, 30, 85, 44, 19, 48, 10, 40, 52, 31, 62, 16, 50,
       98, 64, 60, 99, 90, 87, 87, 53, 67, 63, 83, 53, 59, 43, 21, 44])>

In [73]:
tf.size(E), tf.shape, E.ndim

(<tf.Tensor: shape=(), dtype=int32, numpy=50>,
 <function tensorflow.python.ops.array_ops.shape_v2>,
 1)

In [74]:
#Find the minimum
tf.reduce_min(E)

<tf.Tensor: shape=(), dtype=int64, numpy=2>

In [75]:
#Find the maximim
tf.reduce_max(E)

<tf.Tensor: shape=(), dtype=int64, numpy=99>

In [76]:
#Find the mean
tf.reduce_mean(E)

<tf.Tensor: shape=(), dtype=int64, numpy=54>

In [77]:
#Find the sum
tf.reduce_sum(E)

<tf.Tensor: shape=(), dtype=int64, numpy=2720>

In [78]:
import tensorflow_probability as tfp
#Find the variance
print(tfp.stats.variance(E))
print(tf.math.reduce_variance(tf.cast(E, dtype=tf.float32)))

tf.Tensor(671, shape=(), dtype=int64)
tf.Tensor(671.08, shape=(), dtype=float32)


In [79]:
#Find the standard deviation
tf.math.reduce_std(tf.cast(E, dtype=tf.float32))

<tf.Tensor: shape=(), dtype=float32, numpy=25.905212>

###Find the positional maximum and minimum

In [80]:
#Create a new tensor for finding positional minumim and maximum
tf.random.set_seed(42)
F = tf.random.uniform(shape=[50])
F

<tf.Tensor: shape=(50,), dtype=float32, numpy=
array([0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
       0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
       0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
       0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
       0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
       0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
       0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
       0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
       0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
       0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043],
      dtype=float32)>

In [81]:
#Find the positional maximum
tf.argmax(F)

<tf.Tensor: shape=(), dtype=int64, numpy=42>

In [82]:
#index on our largest value position
F[tf.argmax(F)]

<tf.Tensor: shape=(), dtype=float32, numpy=0.9671384>

In [83]:
#Find the max value of F
tf.reduce_max(F)

<tf.Tensor: shape=(), dtype=float32, numpy=0.9671384>

In [84]:
#Check for equality
tf.reduce_max(F) == F[tf.argmax(F)]

<tf.Tensor: shape=(), dtype=bool, numpy=True>

In [85]:
#Find the positional minimum
tf.argmin(F)

<tf.Tensor: shape=(), dtype=int64, numpy=16>

In [86]:
#index on our smallest value position
F[tf.argmin(F)]

<tf.Tensor: shape=(), dtype=float32, numpy=0.009463668>

In [87]:
#Find the max value of F
tf.reduce_min(F)

<tf.Tensor: shape=(), dtype=float32, numpy=0.009463668>

In [88]:
#Check for equality
tf.reduce_min(F) == F[tf.argmin(F)]

<tf.Tensor: shape=(), dtype=bool, numpy=True>

###Squeezing a tensor (removing all single dimensions)

In [89]:
#Create a tensor to get started
tf.random.set_seed(42)
G = tf.constant(tf.random.uniform(shape=[50]), shape=(1,1,1,1,50))
G

<tf.Tensor: shape=(1, 1, 1, 1, 50), dtype=float32, numpy=
array([[[[[0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
           0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
           0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
           0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
           0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
           0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
           0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
           0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
           0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
           0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043]]]]],
      dtype=float32)>

In [90]:
#print shape and dims
G.shape, G.ndim

(TensorShape([1, 1, 1, 1, 50]), 5)

In [91]:
G_squeezed = tf.squeeze(G)
G_squeezed.shape

TensorShape([50])

###One-hot encoding tensors

In [92]:
#Create a list of indices
some_list = [0,2,3,4] #could be red, green, blue, purple

#One hot encode our list of indices
tf.one_hot(some_list, 4)

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[1., 0., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.],
       [0., 0., 0., 0.]], dtype=float32)>

In [93]:
#Specify costum values for one hot encoding
#on_value for 1
#off_value for 0
tf.one_hot(some_list, depth=4, on_value="i love deep learning", off_value="i love machine learning")

<tf.Tensor: shape=(4, 4), dtype=string, numpy=
array([[b'i love deep learning', b'i love machine learning',
        b'i love machine learning', b'i love machine learning'],
       [b'i love machine learning', b'i love machine learning',
        b'i love deep learning', b'i love machine learning'],
       [b'i love machine learning', b'i love machine learning',
        b'i love machine learning', b'i love deep learning'],
       [b'i love machine learning', b'i love machine learning',
        b'i love machine learning', b'i love machine learning']],
      dtype=object)>

###Squaring, log, square root

In [94]:
H = tf.range(1, 10)
H

<tf.Tensor: shape=(9,), dtype=int32, numpy=array([1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=int32)>

In [95]:
tf.square(H)

<tf.Tensor: shape=(9,), dtype=int32, numpy=array([ 1,  4,  9, 16, 25, 36, 49, 64, 81], dtype=int32)>

In [96]:
#Finde the squareroot
tf.sqrt(tf.cast(H, dtype=tf.float32))

<tf.Tensor: shape=(9,), dtype=float32, numpy=
array([1.       , 1.4142135, 1.7320508, 2.       , 2.236068 , 2.4494898,
       2.6457512, 2.828427 , 3.       ], dtype=float32)>

In [97]:
#Find the log
tf.math.log(tf.cast(H, dtype=tf.float32))

<tf.Tensor: shape=(9,), dtype=float32, numpy=
array([0.       , 0.6931472, 1.0986123, 1.3862944, 1.609438 , 1.7917595,
       1.9459102, 2.0794415, 2.1972246], dtype=float32)>

###Tensors and NumPy

TensorFlow interacts beautifully with NumPy arrays

In [98]:
#Create a tensor directly from a NumPy array
J = tf.constant(np.array([3., 7., 10.]))
J

<tf.Tensor: shape=(3,), dtype=float64, numpy=array([ 3.,  7., 10.])>

In [99]:
#Convert our tensor back to a NumPy array
np.array(J), type(np.array(J))

(array([ 3.,  7., 10.]), numpy.ndarray)

In [100]:
#Convert tensor J to a NumPy array
J.numpy(), type(J.numpy())

(array([ 3.,  7., 10.]), numpy.ndarray)

In [101]:
J = tf.constant([3.])
J.numpy()[0]

3.0

In [102]:
#The default types of each are slightly different
numpy_J = tf.constant([np.array([3., 7., 10.])])
tensor_J = tf.constant([3., 7., 10.])

In [103]:
#Check th datatypes of each
numpy_J.dtype, tensor_J.dtype

(tf.float64, tf.float32)

###Finding access to GPUs

In [105]:
tf.config.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [106]:
!nvidia-smi

Fri Jun 11 16:56:54 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 465.27       Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   73C    P0    33W /  70W |    224MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

###Tensorflow Fundamentals Exercised
* Create a vector, scalar, matrix and tensor with values of your choosing using tf.constant().
* Find the shape, rank and size of the tensors you created in 1.
* Create two tensors containing random values between 0 and 1 with shape [5, 300].
* Multiply the two tensors you created in 3 using matrix multiplication.
* Multiply the two tensors you created in 3 using dot product.
* Create a tensor with random values between 0 and 1 with shape [224, 224, 3].
* Find the min and max values of the tensor you created in 6 along the first axis.
* Created a tensor with random values of shape [1, 224, 224, 3] then squeeze it to change the shape to [224, 224, 3].
* Create a tensor with shape [10] using your own choice of values, then find the index which has the maximum value.
* One-hot encode the tensor you created in 9.